accepted/2025/PrunePackageReference-with-direct-PackageReference.md

```diff
@@ -0,0 +1,264 @@
+The only place the id appears in the assets file is the `project` section under the `packagesToPrune`.
+
+When the NU1510 warning is raised, we are giving the customer a clear signal that they need to remove the PackageReference.
+In these scenarios, what happens to the direct package references is not very critical.
+
+In the partial prune scenarios such as 2b and 2c, pruning directs like transitive could be beneficial in the .NET leg, helping avoid additional audit warnings.
+However, due to the fact that the PackageReference is in other frameworks, the benefit is more minimal, since if the package itself had a vulnerability, the audit warning will be warranted since the customer is using it in a case where it is not being pruned.
+
+#### Should we prune direct PackageReference similarly to transitives?
+
+[Pruning of direct packages](#how-pruning-direct-packagereference-similarly-to-transitive-would-work) would introduce the new concept of a pruned direct package.
+For transitive pruning, the end user would not see anything beyond a reduction in packages downloaded. For direct packages, the PM UI, list & why commands need to updated as well. From a cost/benefit perspective, direct pruning requires a lot of parts to be updated and there is not a big benefit in the core scenario where a package has vulnerabilities (warnings would still be raised since the package is still used).
+Only 5% of .NET SDK based projects are suspected to be multi-targeted, and it would be challenging to get more specificity at this point.
+
+#### Direct PackageReference in the pruning range are treated with `PrivateAssets=all` and `IncludeAssets=none`
+
+If a direct PackageReference is within the pruning range, we can treat with `PrivateAssets=all` and `IncludeAssets=none`.
+
+The direct packages are still resolved. No need to introduce special concepts for direct packages that are not resolved.
+
+`PrivateAssets=all` means that they will not be:
+
+- Transitive project through project - this is pretty inconsequential to the larger story, since as a transitive package, it would be pruned in the referencing project as well (assuming pruning wasn't explicitly disabled).
+- Be part of the resulting package - This is beneficial in scenarios like 2a, 2b and 2c where we can't raise an NU1510 warning due to the risks of conditional PackageReference. We would effectively be removing this package for the user, without them having to apply it themselves.
+Additionally, in single framework scenarios, prunable packages would be removed from the package definition completely, which is effectively NuGet applying the `NU1510` heuristic for the customer. Given that the `NU1510` warning is enabled for `.NET 10` only, this would clean-up packages targeting net9.0 and earlier of unnecessary dependencies.
+
+`IncludeAssets=none` will be used to clearly indicate that the package would not be used at compile or runtime.
```

Comment:
One annoying thing about this is that if the package has Analyzers the SDK will still raise those :-(.  I was we could clean that up and make nuget responsible for raising analyzer items and delete the hack in the SDK.

Reply:
https://github.com/NuGet/Home/issues/6279 is scheduled for @martinrrm and @kfikadu. 

We won't roll it out by default in .NET 10, but we'll soon cover that gap.

Reply:
Thank you 🙇  That's great news.